app/src/main/java/io/supertokens/session/SuperTokens.java

```diff
@@ -83,4 +90,76 @@ public static boolean doesSessionExist(Context context) {
         String idRefreshToken = IdRefreshToken.getToken(context);
         return idRefreshToken != null;
     }
+
+    @SuppressWarnings("unused")
+    public static HttpURLConnection signOut(Context context) throws IOException, IllegalAccessException {
+        if (!doesSessionExist(context)) {
+            return null;
+        }
+
+        URL signOutUrl = new URL(SuperTokens.signOutUrl);
+        HttpURLConnection connection = SuperTokensHttpURLConnection.newRequest(signOutUrl, new SuperTokensHttpURLConnection.PreConnectCallback() {
+            @Override
+            public void doAction(HttpURLConnection con) throws IOException {
+                Map<String, String> customHeaders = SuperTokens.config.customHeaderMapper.getRequestHeaders(CustomHeaderProvider.RequestType.SIGN_OUT);
+                if (customHeaders != null) {
+                    for (Map.Entry<String, String> entry : customHeaders.entrySet()) {
+                        con.setRequestProperty(entry.getKey(), entry.getValue());
+                    }
+                }
+            }
+        });
+
+        int responseCode = connection.getResponseCode();
+
+        if (responseCode == SuperTokens.config.sessionExpiredStatusCode) {
+            // refresh must have already sent session expiry event
+            return null;
+        }
+
+        if (responseCode >= 300) {
+            return connection;
+        }
+
+        return null;
+    }
+
+    public static boolean attemptRefreshingSession(Context context) throws IOException {
+        String preRequestIdRefreshToken = IdRefreshToken.getToken(context);
+        Utils.Unauthorised unauthorisedResponse = SuperTokensHttpURLConnection.onUnauthorisedResponse(preRequestIdRefreshToken, context);
+
+        if (unauthorisedResponse.status == Utils.Unauthorised.UnauthorisedStatus.API_ERROR) {
+            throw unauthorisedResponse.error;
+        }
+
+        return unauthorisedResponse.status == Utils.Unauthorised.UnauthorisedStatus.RETRY;
+    }
+
+    public static String getUserId(Context context) throws JSONException, IOException {
+        JSONObject tokenInfo = FrontToken.getTokenInfo(context);
+        if (tokenInfo == null) {
+            throw new IOException("No session exists");
```

Comment:
throw IllegalStateException. ALso wherever we catch JSONException, we should instead throw IllegalStateException